test/register_manager.cpp

```diff
@@ -0,0 +1,91 @@
+    TreeOne<CqasmV3xVersion> version = cqasm_tree::make<CqasmV3xVersion>("3.0");
+    TreeOne<CqasmV3xBlock> empty_block = cqasm_tree::make<CqasmV3xBlock>();
+    TreeOne<CqasmV3xVariable> qubit_register_of_size_2 = cqasm_tree::make<CqasmV3xVariable>(
+        "qr_of_size_2", cqasm_tree::make<CqasmV3xQubitType>(2));
+    TreeOne<CqasmV3xVariable> qubit_register_of_size_4 = cqasm_tree::make<CqasmV3xVariable>(
+        "qr_of_size_4", cqasm_tree::make<CqasmV3xQubitType>(4));
+    TreeOne<CqasmV3xVariable> too_long_qubit_register = cqasm_tree::make<CqasmV3xVariable>(
+        "too_long_qr", cqasm_tree::make<CqasmV3xQubitType>(qx::config::MAX_QUBIT_NUMBER + 1));
+    TreeOne<CqasmV3xVariable> too_long_bit_register = cqasm_tree::make<CqasmV3xVariable>(
+        "too_long_br", cqasm_tree::make<CqasmV3xBitType>(qx::config::MAX_BIT_NUMBER + 1));
+    TreeOne<CqasmV3xProgram> null_program = TreeOne<CqasmV3xProgram>{};
+    TreeOne<CqasmV3xProgram> program_with_2_qubits = cqasm_tree::make<CqasmV3xProgram>(
+        api_version, version, empty_block, CqasmV3xVariables{ qubit_register_of_size_2 }
+    );
+    TreeOne<CqasmV3xProgram> program_with_4_qubits = cqasm_tree::make<CqasmV3xProgram>(
+        api_version, version, empty_block, CqasmV3xVariables{ qubit_register_of_size_4 }
+    );
+    TreeOne<CqasmV3xProgram> program_with_too_long_qubit_register = cqasm_tree::make<CqasmV3xProgram>(
+        api_version, version, empty_block, CqasmV3xVariables{ too_long_qubit_register }
+    );
+    TreeOne<CqasmV3xProgram> program_with_too_long_bit_register = cqasm_tree::make<CqasmV3xProgram>(
+        api_version, version, empty_block, CqasmV3xVariables{ too_long_bit_register }
+    );
+};
+
+TEST_F(RegisterManagerTest, qubit_register_size_exceeds_maximum_allowed) {
+    EXPECT_THAT(
+        [this]() {
+            [[maybe_unused]] auto qubit_register = QubitRegister{ program_with_too_long_qubit_register };
+        },
+        ::testing::ThrowsMessage<std::runtime_error>(
+            fmt::format("register size exceeds maximum allowed: {} > {}",
+                        qx::config::MAX_QUBIT_NUMBER + 1, qx::config::MAX_QUBIT_NUMBER)
+        )
+    );
+}
+
+TEST_F(RegisterManagerTest, bit_register_size_exceeds_maximum_allowed) {
+    EXPECT_THAT(
+        [this]() {
+            [[maybe_unused]] auto bit_register = BitRegister{ program_with_too_long_bit_register };
+        },
+        ::testing::ThrowsMessage<std::runtime_error>(
+            fmt::format("register size exceeds maximum allowed: {} > {}",
+                        qx::config::MAX_BIT_NUMBER + 1, qx::config::MAX_BIT_NUMBER)
+        )
+    );
+}
+
+TEST_F(RegisterManagerTest, create_instance_with_null_program) {
+    EXPECT_THAT(
+        [this]() {
+            RegisterManager::create_instance(null_program);
+            [[maybe_unused]] const auto &unused = RegisterManager::get_instance();
+        },
+        ::testing::ThrowsMessage<std::runtime_error>("null pointer to program")
```

Comment:
Is this error message informative enough to debug the issue?

Reply:
This exception may well be implemented just as an assert, a check intended to provide some information to the programmer more than to the user. This is, the user of the simulator should never see this message. Then, for the programmer, it is more than descriptive, I would say, mainly because it is the only place where it shows up in the whole codebase.

Reply:
I'm going to remove it though, along with the test, because there is already an assert in the `execute` method, in `simulator.cpp`, line 55. That renders the check in the Register constructor useless:

```
assert(!program.empty())
```

I will change the assert though to provide some more info:

```
assert(!program.empty() && "null pointer to program")
```

Reply:
In the end, I updated the `get_analysis_result` in `simulator.cpp` to also throw a SimulationError if the analyzer returns a null program. I made the code of that function more symmetric. And I updated the error messages. Quite arguably, a null program check would be more an assert than an exception, but it doesn't hurt to implement it as an exception.